File: WebDriverTests_Csharp/Autotest.Core/Autotests.Utilities/WebTestCore/TestSystem/PageLoadCounter.cs
using System;
using Autotests.Utilities.WebTestCore.SystemControls;
using NUnit.Framework;

namespace Autotests.Utilities.WebTestCore.TestSystem
{
    public static class PageLoadCounter
    {
        public static void Reset()
        {
            lastPageLoadCounter = 0;
            cookieName = "pageLoadCounterValue_" + Guid.NewGuid();
        }

        public static void AssertPageNotLoaded()
        {
            var expectedPageLoadCounter = lastPageLoadCounter;
//            Update();
            Assert.AreEqual(expectedPageLoadCounter, lastPageLoadCounter, "Был выполнен неожиданный переход на страницу");
        }

        public static void WaitPageLoaded(int pageLoads = 1, TimeSpan? timeout = null)
        {
//            var pageLoadCounter = lastPageLoadCounter + pageLoads;
//            Waiter.Wait(() => CheckPageLoaded(pageLoadCounter), "Ожидание завершения перехода на страницу", (int)(timeout ?? TimeSpan.FromSeconds(20)).TotalMilliseconds);
            CheckServerException();
//            Assert.AreEqual(pageLoadCounter, lastPageLoadCounter, "Был выполнен неожиданный переход на страницу");
        }

//        public static void InitPageLoadCounterCookie()
//        {
//            if (WebDriverCache.WebDriver.FindCookie(cookieName) == null)
//                WebDriverCache.WebDriver.SetCookie(cookieName, "1");
//        }

//        private static bool CheckPageLoaded(long pageLoadCounter)
//        {
////            Update();
//            if (lastPageLoadCounter >= pageLoadCounter)
//                return true;
//            CheckServerException();
//            return false;
//        }

//        private static void Update()
//        {
//            long.TryParse(WebDriverCache.WebDriver.FindCookie(cookieName), out lastPageLoadCounter);
//        }

        private static void CheckServerException()
        {
            var errorNumber = new StaticText("err-info");
            if (errorNumber.IsPresent)
            {
                errorNumber.SendKeysToBody("h");
                var exceptionText = new StaticText("stackTrace").GetText();
                if (!string.IsNullOrEmpty(exceptionText))
                    throw new Exception("Внутренняя ошибка сервера:" + Environment.NewLine + exceptionText);
            }
        }

        private static long lastPageLoadCounter;
        private static string cookieName;
    }
}

File: WebDriverTests_Java/WebDriverTests/src/test/java/core/systemControls/PageLoadCounter.java
package core.systemControls;


import com.sun.jna.platform.win32.Guid;
import core.webDriver.WebDriverCache;
import org.testng.Assert;

public final class PageLoadCounter
{
    public static void Reset()
    {
        lastPageLoadCounter = 0;
        cookieName = "pageLoadCounterValue_" + new Guid.GUID().toGuidString();
    }

    public static void AssertPageNotLoaded()
    {
        long expectedPageLoadCounter = lastPageLoadCounter;
//        Update();
        Assert.assertEquals(expectedPageLoadCounter, lastPageLoadCounter, "Был выполнен неожиданный переход на страницу");
    }

    private static void Update()
    {
        Long tempRef_lastPageLoadCounter = lastPageLoadCounter;
        lastPageLoadCounter = tempRef_lastPageLoadCounter + Long.parseLong(WebDriverCache.getWebDriver().FindCookie(cookieName));
    }

    public static void InitPageLoadCounterCookie()
    {
        if(WebDriverCache.getWebDriver().FindCookie(cookieName) == null)
            WebDriverCache.getWebDriver().SetCookie(cookieName, "1");
    }
    public static void WaitPageLoaded(int pageLoads)
    {
        WaitPageLoaded(pageLoads, null);
    }

    public static void WaitPageLoaded()
    {
        WaitPageLoaded(1, null);
    }

//    public static void WaitPageLoaded()
//    {
//        WaitPageLoaded(1);
//    }

//     public static void WaitPageLoaded(int pageLoads)
//    {
//        CheckServerException();
//    }


    public static void WaitPageLoaded(int pageLoads, Integer timeout)
    {
        long pageLoadCounter = lastPageLoadCounter + pageLoads;
        Waiter.Wait(() -> CheckPageLoaded(pageLoadCounter), "Ожидание завершения перехода на страницу", (timeout != null) ? timeout : 30000);
        CheckServerException();
        Assert.assertEquals(pageLoadCounter, lastPageLoadCounter, "Был выполнен неожиданный переход на страницу");
    }

    private static boolean CheckPageLoaded(long pageLoadCounter)
    {
        Update();
        if (lastPageLoadCounter >= pageLoadCounter)
        {
            return true;
        }
        CheckServerException();
        return false;
    }

    private static void CheckServerException()
    {
        StaticText errorNumber = new StaticText("err-info");
        if (errorNumber.getIsPresent())
        {
            errorNumber.sendKeysToBody("h");
            String exceptionText = (new StaticText("stackTrace")).getText();
            if (exceptionText.isEmpty())
            {
                throw new RuntimeException("Внутренняя ошибка сервера:" + System.lineSeparator() + exceptionText);
            }
        }
    }

    private static long lastPageLoadCounter;
    private static String cookieName;
}


